### PR TITLE
[Olmo3] different RoPE per layer type

### PR DESCRIPTION
**Pull request by zucchini-nlp**
# What does this PR do?

Reverts back per layer-type RoPE in Olmo which was removed in https://github.com/huggingface/transformers/pull/39847

### CODE REVIEW

File: tests/models/olmo3/test_modeling_olmo3.py [hunk 1]
```diff
@@ -214,10 +217,50 @@ def test_model_7b_greedy_generation(self):
...
-        self.assertEqual(EXPECTED_TEXT_COMPLETION, text)
+        self.assertEqual(expectations.get_expectation(), text)
+
+    def test_real_model_7b_greedy_generation(self):
+        expectations = Expectations(
+            {
+                ("cuda", None): 'system\nYou are a helpful function-calling AI assistant. You do not currently have access to any functions. <functions></functions>\nuser\nWho would win in a fight - a dinosaur or a cow named Moo Moo?\nassistant\nThis is a fun and imaginative question! Let’s break it down:\n\n### 1. **A Dinosaur (General Case)**\nDinosaurs were a huge and diverse group, spanning from tiny feathered raptors to massive sauropods like *Brachiosaurus* or *Tyrannosaurus rex',
+            }
+        )  # fmt: skip
+
```
Comment (zucchini-nlp) [inline tests/models/olmo3/test_modeling_olmo3.py]: there were no integration tests with official ckpt, somehow it redirects to someone's personal repo 🫠 

I added tests with ckpt that are supposed to use rope scaling, and we should see a difference now . Adding test for long seq beyond sliding window in a sec

Comment (zucchini-nlp) [reply tests/models/olmo3/test_modeling_olmo3.py]: i didn't delete existing slow tests, not sure if that repo is supposed to be tested. LMK if you think we can delete everything to not waste resources on running them

File: tests/causal_lm_tester.py [hunk 2]
```diff
@@ -628,7 +628,7 @@ def _set_config_rope_params(config: PreTrainedConfig, rope_params: dict) -> bool
     config.rope_parameters = getattr(config, "rope_parameters", {}) or {}
     config.rope_parameters.update(rope_params)
 
-    if any(name in config.__class__.__name__.lower() for name in ["gemma3", "modernbert"]):
+    if any(name in config.__class__.__name__.lower() for name in ["gemma3", "modernbert", "olmo3"]):
```
Comment (zucchini-nlp) [inline tests/causal_lm_tester.py:631]: added a class name for now, a better heuristic coming in https://github.com/huggingface/transformers/pull/46912

**Comment by zucchini-nlp** [comment]
run-slow: olmo3, olmo_hybrid

File: src/transformers/models/olmo_hybrid/modeling_olmo_hybrid.py [hunk 3]
```diff
@@ -958,6 +958,7 @@ def __init__(self, config: OlmoHybridConfig):
             ]
         )
         self.norm = OlmoHybridRMSNorm(config.hidden_size, eps=config.rms_norm_eps)
+        # Released ckpt don't use any ROPE and have  it set to `None`
         self.rotary_emb = (
             OlmoHybridRotaryEmbedding(config=config)
             if getattr(config, "rope_parameters", None) is not None
```
Comment (zucchini-nlp) [inline src/transformers/models/olmo_hybrid/modeling_olmo_hybrid.py:963]: ig we can't delete the module, since some users might have added rope in non-official ckpts

**Comment by zucchini-nlp** [comment]
run-slow: olmo3, olmo_hybrid

File: tests/causal_lm_tester.py (hunk 2 above)
Comment (vasqu) [reply tests/causal_lm_tester.py:631]: Hmm there are more models with per layer type rope, e.g. one that comes to mind is laguna. But I guess with the tests refactors all those overrides would be handled natively either way. Possibly a simple check whether the rope is nested(?)

File: tests/models/olmo3/test_modeling_olmo3.py [hunk 4]
```diff
@@ -203,9 +223,12 @@ def test_model_7b_logits(self):
         EXPECTED_SLICE = torch.tensor(expectations.get_expectation(), device=torch_device)
         torch.testing.assert_close(out[0, 0, :30], EXPECTED_SLICE, rtol=1e-2, atol=1e-2)
 
-    @slow
     def test_model_7b_greedy_generation(self):
-        EXPECTED_TEXT_COMPLETION = """Simply put, the theory of relativity states that 1) the laws of physics are the same for all observers, and 2) the speed of light is the same for all observers. The first part of the theory is called the principle of relativity, and the second part is called the principle of the constancy of the speed of light. The theory of rel"""
+        expectations = Expectations(
+            {
+                ("cuda", None): """Simply put, the theory of relativity states that 1) the laws of physics are the same for all observers, and 2) the speed of light is the same for all observers. The first part of the theory is called the principle of relativity, and the second part is called the principle of the constancy of the speed of light. The theory of rel""",
```
Comment (vasqu) [inline tests/models/olmo3/test_modeling_olmo3.py:233]: ok same but wrapped around expectations, just to be sure. on a train and it's wonky here 😆

File: tests/models/olmo3/test_modeling_olmo3.py (hunk 1 above)
Comment (vasqu) [reply tests/models/olmo3/test_modeling_olmo3.py]: imo, best to keep but move/copy to internal testing. personal repos is not so nice

no need to save on resources, rather have something non broken

File: tests/models/olmo3/test_modeling_olmo3.py [hunk 5]
```diff
@@ -214,10 +237,78 @@ def test_model_7b_greedy_generation(self):
...
+
+    def test_generate_beyond_sliding_window(self):
+        expectations = Expectations(
+            {
+                ("cuda", None): """It looks like you've pasted a very lengthy and repetitive list of "This is a nice place""",
+            }
+        )  # fmt: skip
+
+        tokenizer = AutoTokenizer.from_pretrained("allenai/Olmo-3-7B-Instruct")
+        model = Olmo3ForCausalLM.from_pretrained("allenai/Olmo-3-7B-Instruct", device_map="auto")
```
Comment (vasqu) [inline tests/models/olmo3/test_modeling_olmo3.py]: nit could do these all on setup instead of recreating each time

File: tests/models/olmo3/test_modeling_olmo3.py [hunk 6]
```diff
@@ -62,47 +74,49 @@ class Olmo3ModelTest(CausalLMModelTest, unittest.TestCase):
...
+            model = model_class(config)
+            model.save_pretrained(tmp_dir, save_original_format=True)
+
+            _init_distributed(tp=self.tensor_parallel_size)(_test_tp_generation_quantized_impl)(
+                tmp_dir, model_class, max_new_tokens
+            )
 
     def test_model_rope_scaling_frequencies(self):
         """Tests the frequency properties of the different RoPE scaling types on the model RoPE layer."""
+        # Gemma3n has different RoPE configs per layer type
```
Comment (vasqu) [inline tests/models/olmo3/test_modeling_olmo3.py]:
```suggestion
        # Olmo3 has different RoPE configs per layer type
```
:D

File: tests/models/olmo3/test_modeling_olmo3.py [hunk 7]
```diff
@@ -62,47 +74,49 @@ class Olmo3ModelTest(CausalLMModelTest, unittest.TestCase):
     # used in `test_torch_compile_for_training`
     _torch_compile_train_cls = Olmo3ForCausalLM if is_torch_available() else None
 
-    @parameterized.expand([("linear",), ("dynamic",), ("yarn",)])
-    def test_model_rope_scaling_from_config(self, scaling_type):
-        config, _ = self.model_tester.prepare_config_and_inputs_for_common()
+    @is_tensor_parallel_test
+    def test_tp_generation_quantized(self):
+        # If model uses rope-theta 50k (default value), the test fails
```
Comment (vasqu) [inline tests/models/olmo3/test_modeling_olmo3.py:79]: that is surprising, wondering whether it also affects other models 👀

File: src/transformers/models/olmo_hybrid/modular_olmo_hybrid.py [hunk 8]
```diff
@@ -716,6 +713,7 @@ def __init__(self, config: OlmoHybridConfig):
                 for layer_idx in range(config.num_hidden_layers)
             ]
         )
+        # Released ckpt don't use any ROPE and have  it set to `None`
```
Comment (vasqu) [inline src/transformers/models/olmo_hybrid/modular_olmo_hybrid.py:715]: 😢

File: src/transformers/models/olmo_hybrid/modeling_olmo_hybrid.py (hunk 3 above)
Comment (vasqu) [reply src/transformers/models/olmo_hybrid/modeling_olmo_hybrid.py:963]: yep rather keep it now, that's why it's hard when official ckpts release after integration :/

File: src/transformers/models/olmo3/modeling_olmo3.py [hunk 9]
```diff
@@ -317,18 +329,22 @@ def compute_default_rope_parameters(
...
         with maybe_autocast(device_type=device_type, enabled=False):  # Force float32
             freqs = (inv_freq_expanded.float() @ position_ids_expanded.float()).transpose(1, 2)
             emb = torch.cat((freqs, freqs), dim=-1)
-            cos = emb.cos() * self.attention_scaling
-            sin = emb.sin() * self.attention_scaling
+            cos = emb.cos() * attention_scaling
+            sin = emb.sin() * attention_scaling
 
-        return cos.to(dtype=x.dtype), sin.to(dtype=x.dtype)
+        return cos, sin
```
Comment (vasqu) [inline src/transformers/models/olmo3/modeling_olmo3.py:347]: are we sure this change was correct? looks like they are downcasting here?

File: src/transformers/models/olmo3/configuration_olmo3.py [hunk 10]
```diff
@@ -97,5 +98,32 @@ def __post_init__(self, **kwargs):
             self.num_key_value_heads = self.num_attention_heads
         super().__post_init__(**kwargs)
 
+    def convert_rope_params_to_dict(self, **kwargs):
```
Comment (vasqu) [inline src/transformers/models/olmo3/configuration_olmo3.py:101]: not on you but we should avoid it whenever we can

**Review by vasqu — APPROVED** [review]
I'm only a bit hesitant re that non downcasting on the rope of olmo3 - any source for that?

But other than that agree with most points, just nits/smaller comments

File: tests/causal_lm_tester.py (hunk 2 above)
Comment (zucchini-nlp) [reply tests/causal_lm_tester.py:631]: yeah, we might have a few more added, havent been checking new releases recently. As it is not related to this PR , I'm working on it

File: tests/models/olmo3/test_modeling_olmo3.py (hunk 4 above)
Comment (zucchini-nlp) [reply tests/models/olmo3/test_modeling_olmo3.py:233]: yes, same, just prettier

File: tests/models/olmo3/test_modeling_olmo3.py (hunk 7 above)
Comment (zucchini-nlp) [reply tests/models/olmo3/test_modeling_olmo3.py:79]: ikr, very weird but I didn't want to dig yet. Most models init with 10k by default except for really weird ones, I will check a bit later and open an issue/another PR if necessary

File: src/transformers/models/olmo3/modeling_olmo3.py (hunk 9 above)
Comment (zucchini-nlp) [reply src/transformers/models/olmo3/modeling_olmo3.py:347]: yeah, they were accidentally reverted in above huge PR 😢

File: tests/models/olmo3/test_modeling_olmo3.py (hunk 7 above)
Comment (zucchini-nlp) [reply tests/models/olmo3/test_modeling_olmo3.py:79]: hmm seems like it is an error accumulation from QK norm over the whole hidden-dim and higher theta, I'll just leave it there as there are more models skipping this particular test with `todo: investigate`
maybe we'd need to raise tols for certain cases

**Comment by Cyrilvallez** [comment]
cc @zucchini-nlp @vasqu why are the `cos`/`sin` not downcasted at the end here like any other Rope module? Any reason? This means any later ops will be upcasted in the model...

**Comment by zucchini-nlp** [comment]
@Cyrilvallez yeas, it is done on purpose to follow official ref impl. QK are then recasted back to model dtype after rotating

https://github.com/huggingface/transformers/blob/dff4572dfa4bfa9f00cc8414e4b84877552fefe9/src/transformers/models/olmo3/modeling_olmo3.py#L121-L126

**Comment by Cyrilvallez** [comment]
All right, thanks for confirming!
